jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java

```diff
@@ -341,32 +342,42 @@ public ContextHandler createContextHandler(final App app) throws Exception
 
             // prepare properties
             Map<String, String> properties = new HashMap<>();
+
+            //add in properties from start mechanism
             properties.putAll(getProperties());
-            properties.putAll(app.getProperties());
 
-            // Handle a context XML file
-            if (FileID.isXml(path))
+            Object context = null;
```

Comment:
I think for this to be fully flexible, we need to ability to specify in the webapps properties, a specific ContextHandler type to be created, else it may be created as a generic env type by the environment.xml below and then the specific context.xml might fail assuming an extended type.

```suggestion
            Object context = null;
            String contextHandlerClass = properties.get(Deployable.CONTEXT_HANDLER_CLASS);
            if (contextHandlerClass != null)
                context = Class.forName(contextHandlerClass).getDeclaredConstructor().newInstance();
```